eth2util/rlp/rlp.go

```diff
@@ -0,0 +1,192 @@
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package rlp implements the simple byte slice and lists of byte slices encoding/decoding using
+// recursive length prefix encoding scheme as per spec
+// https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/.
+package rlp
+
+import (
+	"encoding/binary"
+
+	"github.com/obolnetwork/charon/app/errors"
+)
+
+// DecodeBytesList returns the list of byte slices contained in the given RLP encoded byte slice.
+func DecodeBytesList(input []byte) ([][]byte, error) {
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	offset, length, err := decodeLength(input)
+	if err != nil {
+		return nil, err
+	}
+
+	if offset+length > len(input) {
+		return nil, errors.New("input too short")
+	}
+
+	var items [][]byte
+	for i := offset; i < offset+length; {
+		itemOffset, itemLength, err := decodeLength(input[i:])
+		if err != nil {
+			return nil, err
+		}
+
+		start := i + itemOffset
+		end := i + itemOffset + itemLength
+
+		if end > len(input) {
+			return nil, errors.New("input too short")
+		}
+
+		items = append(items, input[start:end])
+
+		i = end
+	}
+
+	return items, nil
+}
+
+// EncodeBytesList returns a byte slice containing the RLP encoding of the given list of byte slices.
+func EncodeBytesList(items [][]byte) ([]byte, error) {
+	var output []byte
+
+	for _, item := range items {
+		b, err := EncodeBytes(item)
+		if err != nil {
+			return nil, err
+		}
+
+		output = append(output, b...)
+	}
+
+	length, err := encodeLength(len(output), 0xc0)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(length, output...), nil
+}
+
+// DecodeBytes returns the byte slices contained in the given RLP encoded byte slice.
+func DecodeBytes(input []byte) ([]byte, error) {
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	offset, length, err := decodeLength(input)
+	if err != nil {
+		return nil, err
+	}
+
+	if offset+length > len(input) {
+		return nil, errors.New("input too short")
+	}
+
+	return input[offset : offset+length], nil
+}
+
+// EncodeBytes returns a byte slice containing the RLP encoding of the given byte slice.
+func EncodeBytes(item []byte) ([]byte, error) {
+	if len(item) == 1 && item[0] < 0x80 {
+		return item, nil
+	}
+
+	length, err := encodeLength(len(item), 0x80)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(length, item...), nil
+}
+
+// decodeLength returns the length of the RLP encoding prefix (offset) and the length of the encoded byte slice.
```

Comment:
any specific reason to have named return values?

Reply:
since it isn't obvious from the types what is returned

e.g., what is the first in and what is the second int?
```
func decodeLength(item []byte) (int, int, error)
```